### ISSUE TITLE
Possible bug in sign function

### ISSUE THREAD

**Issue by HackerzWorkshop** [fchollet/keras#4958]
Hello, I'm trying to build a custom loss function that's using indicators. It seems that the sign function is broken and returns values other than -1,0,+1.

For example, the following function when used as a metric returns arbitrary values between 0 and 1.

def zubby: return 0.5 * (1 + K.sign(K.mean(y_pred-y_true)))

I'm using Python 3.5.2 with Keras 1.2.0 and TensorFlow 0.12.0 on 64-bit Windows 10.

Here's an example output:

batch_size= 5000  nb_epoch= 500
Train on 701728 samples, validate on 431685 samples
Epoch 1/500
6s - loss: 0.0485 - zubby: 0.5226 - val_loss: 0.0357 - val_zubby: 0.0000e+00
Epoch 2/500
5s - loss: 0.0240 - zubby: 0.4560 - val_loss: 0.0384 - val_zubby: 0.0000e+00
Epoch 3/500
5s - loss: 0.0209 - zubby: 0.4941 - val_loss: 0.0421 - val_zubby: 0.0000e+00
Epoch 4/500
5s - loss: 0.0191 - zubby: 0.4774 - val_loss: 0.0390 - val_zubby: 0.0000e+00
Epoch 5/500
5s - loss: 0.0181 - zubby: 0.4727 - val_loss: 0.0404 - val_zubby: 0.0000e+00
Epoch 6/500
5s - loss: 0.0174 - zubby: 0.4988 - val_loss: 0.0417 - val_zubby: 0.0000e+00
Epoch 7/500
5s - loss: 0.0167 - zubby: 0.4870 - val_loss: 0.0382 - val_zubby: 0.0000e+00
Epoch 8/500
5s - loss: 0.0163 - zubby: 0.4988 - val_loss: 0.0364 - val_zubby: 0.0116
Epoch 9/500
5s - loss: 0.0159 - zubby: 0.4988 - val_loss: 0.0422 - val_zubby: 0.0000e+00
Epoch 10/500
5s - loss: 0.0156 - zubby: 0.4703 - val_loss: 0.0394 - val_zubby: 0.0000e+00
Epoch 11/500
5s - loss: 0.0152 - zubby: 0.4703 - val_loss: 0.0408 - val_zubby: 0.0000e+00
Epoch 12/500
5s - loss: 0.0150 - zubby: 0.4870 - val_loss: 0.0389 - val_zubby: 0.0000e+00
Epoch 13/500
5s - loss: 0.0147 - zubby: 0.5725 - val_loss: 0.0407 - val_zubby: 0.0000e+00
Epoch 14/500
5s - loss: 0.0144 - zubby: 0.5130 - val_loss: 0.0398 - val_zubby: 0.0116
Epoch 15/500
5s - loss: 0.0143 - zubby: 0.4988 - val_loss: 0.0404 - val_zubby: 0.0000e+00
Epoch 16/500
5s - loss: 0.0140 - zubby: 0.4845 - val_loss: 0.0407 - val_zubby: 0.1042
Epoch 17/500
5s - loss: 0.0139 - zubby: 0.4845 - val_loss: 0.0398 - val_zubby: 0.0116
Epoch 18/500
5s - loss: 0.0137 - zubby: 0.4916 - val_loss: 0.0408 - val_zubby: 0.0000e+00
Epoch 19/500
5s - loss: 0.0135 - zubby: 0.4133 - val_loss: 0.0407 - val_zubby: 0.3514
Epoch 20/500
5s - loss: 0.0133 - zubby: 0.5012 - val_loss: 0.0403 - val_zubby: 0.1892
Epoch 21/500
5s - loss: 0.0133 - zubby: 0.4941 - val_loss: 0.0383 - val_zubby: 0.5135
Epoch 22/500
5s - loss: 0.0131 - zubby: 0.4941 - val_loss: 0.0408 - val_zubby: 0.2240

**Comment by kgrm** [comment]
That's not the case. The metrics displayed during training are averaged over the entire training / validation sets. To verify the correctness of your function, you should test it on individual inputs.
